### PR TITLE
 [DR] Track replication status

### PR DESCRIPTION
**Pull request by snalli**
This PR adds a trace message to track replication status from cloud to on-prem server.

### CODE REVIEW

File: ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java [hunk 1]
```diff
@@ -133,6 +135,10 @@ public class ReplicaThread implements Runnable {
   // This is used in the test cases
   private Map<DataNodeId, List<ExchangeMetadataResponse>> exchangeMetadataResponsesInEachCycle = null;
 
+  public static final String REPLICA_STATUS_FILE = "onPremReplicaCheckStatus";
+  public static final String RECOVERY_STATUS_FILE = "cloudReplicaRecoveryStatus";
+  protected final BackupCheckerFileManager fileManager;
```
Comment (ankagrawal) [inline ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java:140]: `ReplicaThread` is a generic class used by all the use cases. Why does it need to have `BackupCheckerFileManager`?

File: ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java [hunk 2]
```diff
@@ -198,6 +204,13 @@ public ReplicaThread(String threadName, FindTokenHelper findTokenHelper, Cluster
     }
     this.maxReplicaCountPerRequest = replicationConfig.replicationMaxPartitionCountPerRequest;
     this.leaderBasedReplicationAdmin = leaderBasedReplicationAdmin;
+    try {
+      fileManager = Utils.getObj(replicationConfig.backupCheckFileManagerType, replicationConfig, metricRegistry);
```
Comment (ankagrawal) [inline ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java:208]: Again backup checker specific code here.

**Review by ankagrawal — CHANGES_REQUESTED** [review]
Why are we moving backup checker specific stuff in ReplicaThread?

File: ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java (hunk 1 above)
Comment (JingQianCloud) [reply ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java:140]: Same question. We'd better not touch Replication code. We don't need log for replication thread.

File: ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java (hunk 2 above)
Comment (snalli) [reply ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java:208]: The default backup checker filemgr is empty

File: ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java (hunk 1 above)
Comment (snalli) [reply ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java:140]: 1/ Not touching the replication code, its not logging anything. Default filemgr is empty methods.
2/ I want to log replication status from VCR to server in the DR node and this is the easiest way to do that. I have attempted creating a RecoveryThread here https://github.com/linkedin/ambry/pull/2359
And it seemed like way too much just to log recovery status.
3/ There are only two use cases of ReplicaThread : One to replicate between servers, and one for Recovery from cloud. This code does not interfere with either and is controlled by configs that I will use on DR node.

Comment (snalli) [reply ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java:140]: If you have an easier/shorter or quicker way to do it, lmk.

File: ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java [hunk 3]
```diff
@@ -247,12 +260,60 @@ NetworkClient getNetworkClient() {
...
+
   /**
    * Logs replication progress of local node against some remote node
    * @param remoteReplicaInfo remote replica information
+   * @param exchangeMetadataResponse metadata information from remote node
    */
   protected void logReplicationStatus(RemoteReplicaInfo remoteReplicaInfo,
       ExchangeMetadataResponse exchangeMetadataResponse) {
+    if (replicationConfig.backupCheckFileManagerType.equals(replicationConfig.DEFAULT_BACKUP_CHECKER_FILE_MANAGER)) {
+      // optimization to not affect replication in servers in common case
```
Comment (snalli) [inline ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java:295]: @JingQianCloud @ankagrawal Please take look here. This returns empty. So nothing is logged in common case.

File: ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java [hunk 4]
```diff
@@ -247,12 +260,60 @@ NetworkClient getNetworkClient() {
     return networkClient;
   }
 
+  /**
+   * Returns a concatenated file path
+   * @param remoteReplicaInfo Info about remote replica
+   * @param fileName Name of file to write text to
+   * @return Returns a concatenated file path
+   */
+  protected String getFilePath(RemoteReplicaInfo remoteReplicaInfo, String fileName) {
```
Comment (snalli) [inline ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java:269]: @JingQianCloud @ankagrawal this is not called anywhere in replication code.

**Review by JingQianCloud — APPROVED** [review]

**Review by ankagrawal — APPROVED** [review]

**Pull request merged**